src/tree/gpu_hist/evaluator.cu

```diff
@@ -64,33 +76,33 @@ common::Span<bst_feature_t const> GPUHistEvaluator<GradientSumT>::SortHistogram(
     TreeEvaluator::SplitEvaluator<GPUTrainingParam> evaluator) {
   dh::XGBDeviceAllocator<char> alloc;
   auto sorted_idx = this->SortedIdx(left);
+  auto d_fidx = dh::ToSpan(this->fidx_);
   dh::Iota(sorted_idx);
   // sort 2 nodes and all the features at the same time, disregarding colmun sampling.
-  thrust::stable_sort(
-      thrust::cuda::par(alloc), dh::tbegin(sorted_idx), dh::tend(sorted_idx),
-      [evaluator, left, right] XGBOOST_DEVICE(size_t l, size_t r) {
-        auto l_is_left = l < left.feature_values.size();
-        auto r_is_left = r < left.feature_values.size();
-        if (l_is_left != r_is_left) {
-          return l_is_left;  // not the same node
-        }
+  auto comp = [=] XGBOOST_DEVICE(size_t l, size_t r) {
+    auto l_is_left = l < left.feature_values.size();
+    auto r_is_left = r < left.feature_values.size();
+    if (l_is_left != r_is_left) {
+      return l_is_left;  // not the same node
+    }
 
-        auto const &input = l_is_left ? left : right;
-        l -= (l_is_left ? 0 : input.feature_values.size());
-        r -= (r_is_left ? 0 : input.feature_values.size());
+    auto const &input = l_is_left ? left : right;
+    l -= (l_is_left ? 0 : input.feature_values.size());
+    r -= (r_is_left ? 0 : input.feature_values.size());
 
-        auto lfidx = dh::SegmentId(input.feature_segments, l);
-        auto rfidx = dh::SegmentId(input.feature_segments, r);
-        if (lfidx != rfidx) {
-          return lfidx < rfidx;  // not the same feature
-        }
-        if (common::IsCat(input.feature_types, lfidx)) {
-          auto lw = evaluator.CalcWeightCat(input.param, input.gradient_histogram[l]);
-          auto rw = evaluator.CalcWeightCat(input.param, input.gradient_histogram[r]);
-          return lw < rw;
-        }
-        return l < r;
-      });
+    auto lfidx = d_fidx[l];
+    auto rfidx = d_fidx[r];
+    if (lfidx != rfidx) {
+      return lfidx < rfidx;  // not the same feature
+    }
+    if (common::IsCat(input.feature_types, lfidx)) {
+      auto lw = evaluator.CalcWeightCat(input.param, input.gradient_histogram[l]);
+      auto rw = evaluator.CalcWeightCat(input.param, input.gradient_histogram[r]);
+      return lw < rw;
+    }
+    return l < r;
+  };
+  thrust::stable_sort(thrust::cuda::par(alloc), dh::tbegin(sorted_idx), dh::tend(sorted_idx), comp);
```

Comment:
I wonder if this sort is slow because the comparator is repeatedly performing slow gather memory accesses and evaluating CalcWeightCat? This comparator is evaluated nlogn times. What if you pre-calculate these in a tuple (e.g. by using a transformer iterator as input to sort) and allow thrust to sort the tuples of precaculated values?

Reply:
Yes, that's incredibly helpful!  Will close this PR now and submit a new one.